core/primitives/src/shard_layout.rs

```diff
@@ -444,6 +476,14 @@ impl ShardLayout {
     pub fn shard_uids(&self) -> impl Iterator<Item = ShardUId> + '_ {
         self.shard_ids().map(|shard_id| ShardUId::from_shard_id_and_layout(shard_id, self))
     }
+
+    pub fn get_shard_index(&self, shard_id: ShardId) -> usize {
```

Comment:
nit: we should say this returns a index from 0..num_shards, with stable ordering

btw will there be guarantees about the index of a shard `S` when shard_layout changes and `S` is in the old and new layout?

Reply:
Thanks, I will add comments and implementation later. I just want to make sure that this approach works fine first. 

> btw will there be guarantees about the index of a shard S when shard_layout changes and S is in the old and new layout?

Good question and it's the opposite. The shard indices will change for almost all shards during resharding. It is ok though because the shard index is maintained in the shard layout. The users should make sure to use the correct shard layout when getting the shard index - but again I think this expectation is reasonable and quite common.